### ISSUE TITLE
Performance of cypher.core.Record

### ISSUE THREAD

**Issue by mpreusse** [nigelsmall/py2neo#403]
The Record object is very convenient to work with. But I found that it is sometimes slow compared to built in types. Obviously this is only an issue when you have very long RecordLists.

I have a query that returns ~500.000 nodes. Iterating over the RecordList and retrieving results took longer than I expected.

As a benchmark example, I converted a Record to a list, which touches every element:

    list(some_record)

Converting a record with three elements to a list takes ~10 us (feels slow):

	In [40]: type(r)
	Out[40]: py2neo.cypher.core.Record

	In [41]: %timeit l = list(r)
	The slowest run took 10.87 times longer than the fastest. 
    This could mean that an intermediate result is being cached
	100000 loops, best of 3: 10.3 µs per loop

Converting e.g. a tuple to a list is ~30 times faster:

	In [44]: t = tuple([1, 2, 3])

	In [45]: %timeit list(t)
	The slowest run took 24.89 times longer than the fastest. 
	This could mean that an intermediate result is being cached
	1000000 loops, best of 3: 316 ns per loop

I am wondering where these differences come from. Is it possible to return query results in native data types instead of Record objects? This is less convenient but may be faster when you run many large queries. A list of tuples for example would perform much better when iterating results.

**Comment by nigelsmall** [comment]
To investigate.

**Comment by nigelsmall** [comment]
I'm unclear why you are comparing times for converting Records and tuples into lists. The typical usage pattern for a Record is to extract fields by name or index; it's merely intended to act as a container for these fields.

**Comment by mpreusse** [comment]
I think if you convert a Record to a list it's similar to doing for ... in on the Record or getting the individual columns with myrecord[0], myrecord[1].

I ran a couple of queries that returned 500K to 10M rows, iterated over the Records in the RecordList and did something with the values. It was a bit slow, so I tried to figure out why.

I guess the only point here is that it would be faster to iterate over a list containing tuples with the result values. It's maybe not necessary at all because it's only an issue if you regularly deal with very large result sets.

Anyways, I thought it might be worth pointing out ;)
